tripy/examples/diffusion/model.py

```diff
@@ -0,0 +1,775 @@
+        return self.decoder(latent)
+
+
+# Used for UNet, not to be confused with ResnetBlock, called ResnetBlock2D in HF diffusers
+class ResBlock(tp.Module):
+    def __init__(self, channels, emb_channels, out_channels):
+        self.norm1 = tp.GroupNorm(32, channels)
+        self.conv1 = tp.Conv(channels, out_channels, (3, 3), padding=((1, 1), (1, 1)))
+        self.time_emb_proj = tp.Linear(emb_channels, out_channels)
+        self.norm2 = tp.GroupNorm(32, out_channels)
+        self.conv2 = tp.Conv(out_channels, out_channels, (3, 3), padding=((1, 1), (1, 1)))
+        self.nonlinearity = tp.silu
+        self.conv_shortcut = tp.Conv(channels, out_channels, (1, 1)) if channels != out_channels else lambda x: x
+
+    def __call__(self, x, emb):
+        h = self.conv1(self.nonlinearity(self.norm1(x)))
+        emb_out = self.time_emb_proj(self.nonlinearity(emb))
+        target_shape = tp.concatenate([emb_out.shape, one_shape, one_shape], dim=0)
+        # TODO: #228: WAR to prevent computing output rank in infer_rank for reshape
+        target_shape.trace_tensor.shape = (emb_out.rank + 2,)
+        h = h + tp.reshape(emb_out, target_shape)
+        h = self.conv2(self.nonlinearity(self.norm2(h)))
+        ret = self.conv_shortcut(x) + h
+        return ret
+
+
+class CrossAttention(tp.Module):
+    def __init__(self, query_dim, context_dim, n_heads, d_head):
+        self.to_q = tp.Linear(query_dim, n_heads * d_head, bias=False)
+        self.to_k = tp.Linear(context_dim, n_heads * d_head, bias=False)
+        self.to_v = tp.Linear(context_dim, n_heads * d_head, bias=False)
+        self.num_heads = n_heads
+        self.head_size = d_head
+        self.to_out = [tp.Linear(n_heads * d_head, query_dim)]
+
+    def __call__(self, x, context=None):
+        context = x if context is None else context
+        q, k, v = self.to_q(x), self.to_k(context), self.to_v(context)
+        q, k, v = [
+            tp.transpose(tp.reshape(y, (x.shape[0], -1, self.num_heads, self.head_size)), 1, 2) for y in (q, k, v)
+        ]
+        attention = tp.transpose(scaled_dot_product_attention(q, k, v, embedding_dim=self.head_size), 1, 2)
+        h_ = tp.reshape(attention, (x.shape[0], -1, self.num_heads * self.head_size))
+        out = sequential(h_, self.to_out)
+        return out
+
+
+class GEGLU(tp.Module):
+    def __init__(self, dim_in, dim_out):
+        self.proj = tp.Linear(dim_in, dim_out * 2)
+        self.dim_out = dim_out
+
+    def __call__(self, x):
+        proj = self.proj(x)
+        x, gate = tp.split(proj, 2, proj.rank - 1)  # TODO: allow negative dim in split
+        return x * tp.gelu(gate)
+
+
+class Dummy(tp.Module):
+    def __init__(self):
+        pass
+
+    def __call__(self, x):
+        return x
+
+
+class FeedForward(tp.Module):
+    def __init__(self, dim, mult=4):
+        self.net = [
+            GEGLU(dim, dim * mult),
+            Dummy(),  # needed for weights loading code to work
```

Comment:
I am not sure if I understand why this is required.

Reply:
Also curious. I'm assuming it's to get the list offset right, but what was here in the HF model? Can we just change the weight loading code to account for this?

Reply:
It's to account for a Dropout layer in the HF model and to get the list offset correct, yeah. I can clarify the comment with more details.